examples/pxScene2d/src/pxShaderResource.cpp

```diff
@@ -0,0 +1,1433 @@
+}
+
+void rtShaderResource::prelink()
+{
+  mPosLoc = 0;
+  mUVLoc = 1;
+  glBindAttribLocation(mProgram, mPosLoc, "pos");
+  glBindAttribLocation(mProgram, mUVLoc,  "uv");
+}
+
+void rtShaderResource::postlink()
+{
+  mTimeLoc       = getUniformLocation("u_time");
+  mResolutionLoc = getUniformLocation("u_resolution");
+  mMatrixLoc     = getUniformLocation("amymatrix");
+
+  if(mUniform_map.size() > 0)
+  {
+    for (UniformMapIter_t it  = mUniform_map.begin();
+                          it != mUniform_map.end(); ++it)
+    {
+      (*it).second.loc = getUniformLocation( (*it).second.name );
+    }
+  }//ENDIF
+}
+
+//=====================================================================================================================================
+//
+// GL Setters
+//
+/*static*/ rtError rtShaderResource::fillInt32Vec(int32_t *vec, const rtArrayObject* vals)
+{
+  if(vals && vec)
+  {
+    rtValue count;                  // HACK - WORKAROUND
+    vals->Get("length", &count);    // HACK - WORKAROUND
+
+    int len = count.toUInt32();
+    int  ii = 0;
+
+    for (uint32_t i = 0, l = len; i < l; ++i)
+    {
+      rtValue vv;
+      if(vals->Get(i, &vv) == RT_OK && !vv.isEmpty())
+      {
+        vec[ii++] = vv.toInt32();
+      }
+    }//FOR
+
+    return RT_OK;
+  }
+
+  return RT_FAIL;
+}
+
+/*static*/ rtError rtShaderResource::fillFloatVec(float *vec, const rtArrayObject* vals)
+{
+  if(vals && vec)
+  {
+    rtValue count;                  // HACK - WORKAROUND
+    vals->Get("length", &count);    // HACK - WORKAROUND
+
+    int len = count.toUInt32();
+    int ii  = 0;
+
+    for (uint32_t i = 0, l = len; i < l; ++i)
+    {
+      rtValue vv;
+      if(vals->Get(i, &vv) == RT_OK && !vv.isEmpty())
+      {
+        vec[ii++] = vv.toFloat();
+      }
+    }//FOR
+
+    return RT_OK;
+  }
+
+  return RT_FAIL;
+}
+
+// sampler
+/*static*/ rtError rtShaderResource::bindTexture3(const uniformLoc_t &p)
+{
+  rtImageResource *img = (rtImageResource *) p.value.toObject().getPtr();
+
+  if(img)
+  {
+    pxTextureRef tex = img->getTexture();
+
+    if(tex)
+    {
+      GLuint     tid = tex->getNativeId();
+
+      glActiveTexture(GL_TEXTURE3);
+      glBindTexture(GL_TEXTURE_2D,     tid);
+      glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+      glUniform1i(p.loc, 3);
+
+      return RT_OK;
+    }
+  }
+
+  return RT_FAIL;
+}
+
+// sampler
+/*static*/ rtError rtShaderResource::bindTexture4(const uniformLoc_t &p)
+{
+  rtImageResource *img = (rtImageResource *) p.value.toObject().getPtr();
+
+  if(img)
+  {
+    pxTextureRef tex = img->getTexture();
+
+    if(tex)
+    {
+      GLuint     tid = tex->getNativeId();
+
+      tex->bindTexture();
+
+      glActiveTexture(GL_TEXTURE4);
+      glBindTexture(GL_TEXTURE_2D,     tid);
+      glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+      glUniform1i(p.loc, 4);
+
+      return RT_OK;
+    }
+  }
+
+  return RT_FAIL;
+}
+
+// sampler
+/*static*/ rtError rtShaderResource::bindTexture5(const uniformLoc_t &p)
```

Comment:
Let's combine `bindTexture3`, `bindTexture4`, and `bindTexture5` into single method that takes an additional argument for the texture id.  All the code between these three methods are identical besides the calls to `glActiveTexture()` and `glUniform1i()`, and they are great candidates to combine to a single method.  This will allow for easier code management going forward since only one method update will be needed instead of making the same changes to three separate methods.

Reply:
Like it !

Reply:
Done